src/awkward/_connect/pyarrow.py

```diff
@@ -1114,3 +1114,11 @@ def form_handle_arrow(schema, pass_empty_field=False):
 
         else:
             return out
+
+
+def convert_to_array(layout, args, kwargs):
+    out = ak.operations.to_arrow(layout, extensionarray=False)
+    if args == () and kwargs == {}:
+        return out
+    else:
+        return pyarrow.array(out, *args, **kwargs)
```

Comment:
Your test passes, but I tried what ought to be the same thing and it didn't:

```python
>>> import pyarrow as pa
>>> a = pa.array([1.1, 2.2, 3.3, 4.4, 5.5])
>>> b = pa.array(a, type=pa.float64())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pyarrow/array.pxi", line 317, in pyarrow.lib.array
  File "pyarrow/array.pxi", line 39, in pyarrow.lib._sequence_to_array
  File "pyarrow/error.pxi", line 144, in pyarrow.lib.pyarrow_internal_check_status
  File "pyarrow/error.pxi", line 100, in pyarrow.lib.check_status
pyarrow.lib.ArrowInvalid: Could not convert <pyarrow.DoubleScalar: 1.1> with type pyarrow.lib.DoubleScalar: tried to convert to double
```

(I was going to do a test to find out if wrapping an Arrow array with `pa.array` is zero-copy. It ought to be—I was just checking.)

It might be my pyarrow version (9.0.0): I checkout out this repo and tried to run the test and it failed (same error).

We currently support Arrow versions as low as 7.0.0. This is apparently a bug that they fixed in version $x$ such that $9 < x \le 12$. We could increase that threshold, except that Arrow is an important requirement for basic things like saving and reading files, and other packages have a tendency of putting upper bounds on the Arrow version. (`condastats` wanted to lower it to Arrow 6, which made it tricky to build an environment in which conda stats can actually be analyzed.)

We can't pass `pa.array` arguments into `to_arrow` because `to_arrow` doesn't usually call `pa.array` directly; it uses `pa.Array.from_buffers`, which doesn't take the same arguments. Arrow version $x$ is only needed for the `else` case of this `if`-`else`: maybe check the Arrow version there? If it's less than $x$, raise an error message saying that passing arguments requires Arrow version $x$?

Or maybe that's overkill and we should just explain that it's an Arrow bug, fixed in version $x$, when our users encounter this. It would at least require an Arrow version check in the test, so that we can run the tests with lower versions of Arrow. (And maybe that needs to be added to the CI matrix, not just NumPy minimum version, but Arrow, too.)

Reply:
By the way, my choice would be for the last one: just ensure that our tests work with the Arrow versions we say we support, even if the Arrow version we support has known bugs when some (not all) operations are attempted.

Reply:
Good spot; I assumed that this part of the Arrow API was more stable than this. It looks like only 12.0.0 supports the casting that is performed by the test case in question. I've made it `skipif`

Reply:
Some of the other tests weren't working for me, so I've added a minimum pyarrow to the build matrix. We don't depend on it quite as much as NumPy, but increasingly, and it has more radical changes from one version to the next.

Reply:
That's a good idea. Thanks!

Reply:
Two of these failures are the new tests I was expecting, but some others are from pyarrow string functions. I'm going to test those locally now.

https://github.com/scikit-hep/awkward/actions/runs/5871689180/job/15921684610?pr=2650

Reply:
They're all fine in my pyarrow 9.0.0, but apparently not in the pyarrow 7.0.0 that we say we support. It was good to test!

They're encoding mistakes. I can fix the string functions (in this PR) while you fix the casting.

Reply:
Done!